ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,301 @@
+import {notebooksClusterApi} from 'app/services/notebooks-swagger-fetch-clients';
+import {clusterApi} from 'app/services/swagger-fetch-clients';
+import {isAbortError, reportError} from 'app/utils/errors';
+import {Cluster, ClusterStatus} from 'generated/fetch';
+
+// We're only willing to wait 20 minutes total for a cluster to initialize. After that we return
+// a rejected promise no matter what.
+const DEFAULT_OVERALL_TIMEOUT = 1000 * 60 * 20;
+const DEFAULT_INITIAL_POLLING_DELAY = 2000;
+const DEFAULT_MAX_POLLING_DELAY = 15000;
+// By default, we're willing to retry twice on each of the state-modifying API calls, to allow
+// for transient 500s.
+const DEFAULT_MAX_CREATE_COUNT = 2;
+const DEFAULT_MAX_DELETE_COUNT = 2;
+const DEFAULT_MAX_RESUME_COUNT = 2;
+const DEFAULT_MAX_SERVER_ERROR_COUNT = 5;
```

Comment:
5 seems reasonable for consecutive errors, but might be a bit low for global error count during the polling cycle. May want to make this distinction.